[WIP] Move task icons up to the same line as the task name, saving vertical space

roastchicken opened the pull request:
Especially on smaller screens, space is precious. Having the icons on a separate line takes up vertical space, when there's plenty of room for them to sit to the right of the rest of the task content.

I've also created an issue for this in habitica-android, I'm not sure how separated the codebases are: https://github.com/HabitRPG/habitica-android/issues/1270.

### Changes
Set `position: absolute` on `.icons` in `task.vue`.

Before:
![Skjermbilde 2020-03-13 kl  10 54 17](https://user-images.githubusercontent.com/524709/76610339-0de63580-6519-11ea-8517-ba01bec28ba7.png)

After:
![Skjermbilde 2020-03-13 kl  10 54 10](https://user-images.githubusercontent.com/524709/76610347-10e12600-6519-11ea-8e46-88eb710e5761.png)

A small possible issue is a slight overlap of text, this is the most overlap I was able to acheive:

![Skjermbilde 2020-03-13 kl  10 59 53](https://user-images.githubusercontent.com/524709/76610810-c7450b00-6519-11ea-9e8c-667d57efb241.png)

This could be remedied by increasing `padding-right` on `.task-notes`, or `margin-bottom` on `.task-notes.markdown p`.

----
UUID: 7b9c79ad-ad01-4ef0-a250-01d0e4a73bf3

roastchicken committed:
Move task icons to the same line as the task name

Especially on smaller screens, space is precious. Having the icons on a
separate line takes up vertical space, when there's plenty of room for
them to sit to the right of the rest of the task content.

bottom: 7px; is to mimic the padding-bottom: 7px; on .task-content.

.task-content is given position: relative in order to make it the
containing block for .icons. Without this, .task-content's parent
.d-flex is the containing block which results in .icons overlapping with
.right-control. .task-content is not affected because no offsets (top,
bottom, left, right) are specified.

paglias commented:
thanks @roastchicken ! pinging our designer @Tressley

Tressley commented:
@roastchicken -- The task icons are on their own line to account for the various widths of the task title, task description, checklist items, and task options. Additionally, the layout should be consistent for each task type. Your solution would break down on a To-Do that has a due date.

However, I do feel that there are spacing updates that could be made to shrink the current task height. Here's an example of what a Daily should look like at its smallest height (with icons.)

![Red](https://user-images.githubusercontent.com/2686586/76618534-29444700-64ff-11ea-9e83-d59e07422501.png)

roastchicken committed:
Revert "Move task icons to the same line as the task name"

This reverts commit 4c1be40363e3a19aa0345bba25d730d75ee6e659. Tressley
pointed out that this breaks the To-Do due date icon. I've devised
another solution, preserving the due date.

roastchicken committed:
Move streak, challenge, and tag icons to the right

Especially on smaller screens, space is precious. Having the icons on a
separate line takes up vertical space, when there's plenty of room for
them to sit to the right of the rest of the task content.

Because the diff is quite messy, here's a list of changes I've made:

 - Make .task-clickable-area a flex-container
 - Add an extra div around .task-clickable-area's current contents to
   keep them from being flexed
 - Move .icons-right from being a child of .icons, to being a child of
   .task-clickable-area
 - Make sure .icons-right stays right-aligned and doesn't shrink, as
   well as keeping its icons bottom-aligned
 - Only show .icons if there is a due date, to shave off those 4px of
   margin when it's not needed
 - Change .task-notes margin-bottom to 4px to account for the fact that
   the margin no longer collapses

roastchicken committed:
Move .icons-right out of .task-clickable-area

I realized that the icons were no longer vertically aligned with the due
date, and this became even more obvious when a checklist was added. In
order to keep them aligned, I moved them out of .task-clickable-area and
matched its padding-bottom.

List of changes (messy diff):

 - Remove .d-flex from .task-clickable-area
 - Remove extra div around .task-clickable-area's contents
 - Move .icons-right from being a child of .task-clickable-area, to
   being a child of its parent element

Remaining issues are the border on :hover around the task box, and the options button.

roastchicken commented:
Ah, I hadn't realized that todo-dates are left-aligned.

As I mentioned in https://github.com/HabitRPG/habitica-android/issues/1270, having the icons under the task name makes it too high. There is no reason to have that extra height. This is especially noticable on mobile, where screen real-estate is quite limited.

To alleviate the issue of overlap with various widths, I suggest another solution which I have just commited: moving the icons to the right while keeping the due date on the bottom.

I don't fully see a reason for the layout being consistent between other task types and To-Dos. To-Dos are the only task type (AFAIK) that have an icon left-aligned. @Tressley could you explain your thoughts on that subject?

# Changes

My technical changes are outlined in the commits.

## Desktop

### Before
![desktop_original](https://user-images.githubusercontent.com/524709/76633182-9f1ed180-6544-11ea-9e89-4f0cf8a2fc44.png)

### After
![desktop_move-task-icons](https://user-images.githubusercontent.com/524709/76633199-a5ad4900-6544-11ea-9a55-9ed6dcbf50c6.png)

## Mobile

### Before
![mobile_original](https://user-images.githubusercontent.com/524709/76633211-ab0a9380-6544-11ea-9992-2a64be779b3c.png)

### After
![mobile_move-task-icons](https://user-images.githubusercontent.com/524709/76633228-b1007480-6544-11ea-8675-068b121cb720.png)

IMHO this design looks better.

Some problems remain, namely borders and the options button. I'll continue work on these; I'm not sure why the border is applied to the individual children instead of the parent.

roastchicken committed:
Move <menu-dropdown> to the right in .icons-right

After .icons-right was moved outside of .task-clickable-area, the
options dropdown button was no longer in the top right of the task box
because .icons-right was to the right of it.

List of changes (messy diff):

 - Move <menu-dropdown> from being a child of the parent of .task-title
   to being a child of .icons-right
 - Add a <div> around all of the icons in .icon-right to keep them in a
   row
 - Add padding-top to .icons-right to match .task-clickable-area so the
   dropdown button is in (almost) the same position
 - Remove the margin-top from .icons-right, for the above purpose

<menu-dropdown>'s new position is 1px higher than its previous one,
probably because of some strange layout rules from being inside an extra
div with .task-title.

roastchicken commented:
I've now fixed the border and options button. I've committed the options button, but because the border issue is logically disconnected from this change I've decided to make a separate pull request with it. Once that PR is accepted I'll rebase (and probably squash) this branch off of [develop](/HabitRPG/habitica/tree/develop).

Here's a screenshot of how it looks all together:
![screenshot](https://user-images.githubusercontent.com/524709/76643528-7ce17f80-6555-11ea-97af-f2c1bbf806c8.png)

I welcome any feedback you have @Tressley :)

roastchicken committed:
Nest .icons-right selector in .icons

Also remove flex-grow because we don't want .icons-right to grow now.

Tressley commented:
@roastchicken -- Thanks so much for doing this, but I wish you'd have followed my initial feedback before making these sweeping changes.

> There is no reason to have that extra height.

As I stated before, the reasoning behind our design decision is to ensure the design doesn't break down with the various task types. While I agree this does save space, the most important parts of a task are the title and notes, so they should not compete visually with the icons. Additionally, this allows us to be more efficient in our code.

Again, I have revised task designs that include margin and padding updates that address your vertical height concerns. If you're interested in implementing those designs, I can share the Zeplin artboards with you (just need your email.) However, placing icons and the title on a single line is not the direction we'd like to pursue at this time.

roastchicken commented:
@Tressley
Respectfully, I completely disagree. The design I have implemented does not break any task types, as evidenced by my above screenshots. I don't believe moving the icons slightly up makes them compete with the title or the notes. Further, I don't understand what kind of efficiency you're referring to.

But very well, it seems I won't convice anyone to merge this. Feel free to send the Zeplin artboards to me at the email listed on my GitHub profile.
